include/crocoddyl/core/constraints/residual.hxx

```diff
@@ -0,0 +1,301 @@
+      if (is_rq && is_rv) {
+        data->Hx.row(nh_i) = data->residual->Rx.row(i);
+      } else if (is_rq) {
+        data->Hx.row(nh_i).head(nv) = data->residual->Rx.row(i).head(nv);
+      } else if (is_rv) {
+        data->Hx.row(nh_i).tail(nv) = data->residual->Rx.row(i).tail(nv);
+      }
+      ++nh_i;
+    } else if (constraint_type_(i) == 1) {  // inequality constraint
+      if (is_rq && is_rv) {
+        data->Gx.row(ng_i) = data->residual->Rx.row(i);
+        data->Gx.row(ng_i + 1) = -data->residual->Rx.row(i);
+      } else if (is_rq) {
+        data->Gx.row(ng_i).head(nv) = data->residual->Rx.row(i).head(nv);
+        data->Gx.row(ng_i + 1).head(nv) = -data->residual->Rx.row(i).head(nv);
+      } else if (is_rv) {
+        data->Gx.row(ng_i).tail(nv) = data->residual->Rx.row(i).tail(nv);
+        data->Gx.row(ng_i + 1).tail(nv) = -data->residual->Rx.row(i).tail(nv);
+      }
+      ng_i += 2;
+    } else if (constraint_type_(i) == 2) {  // lower inequality constraint
+      if (is_rq && is_rv) {
+        data->Gx.row(ng_i) = -data->residual->Rx.row(i);
+      } else if (is_rq) {
+        data->Gx.row(ng_i).head(nv) = -data->residual->Rx.row(i).head(nv);
+      } else if (is_rv) {
+        data->Gx.row(ng_i).tail(nv) = -data->residual->Rx.row(i).tail(nv);
+      }
+      ++ng_i;
+    } else if (constraint_type_(i) == 3) {
+      if (is_rq && is_rv) {
+        data->Gx.row(ng_i) = data->residual->Rx.row(i);
+      } else if (is_rq) {
+        data->Gx.row(ng_i).head(nv) = data->residual->Rx.row(i).head(nv);
+      } else if (is_rv) {
+        data->Gx.row(ng_i).tail(nv) = data->residual->Rx.row(i).tail(nv);
+      }
+      ++ng_i;
+    }
+  }
+}
+
+template <typename Scalar>
+boost::shared_ptr<ConstraintDataAbstractTpl<Scalar> > ConstraintModelResidualTpl<Scalar>::createData(
+    DataCollectorAbstract* const data) {
+  return boost::allocate_shared<Data>(Eigen::aligned_allocator<Data>(), this, data);
+}
+
+template <typename Scalar>
+const typename MathBaseTpl<Scalar>::VectorXs& ConstraintModelResidualTpl<Scalar>::get_ub() const {
+  return ub_;
+}
+
+template <typename Scalar>
+const typename MathBaseTpl<Scalar>::VectorXs& ConstraintModelResidualTpl<Scalar>::get_lb() const {
+  return lb_;
+}
+
+template <typename Scalar>
+void ConstraintModelResidualTpl<Scalar>::update_bounds(const VectorXs& lower, const VectorXs& upper) {
+  if (((upper - lower).array() < 0.).any()) {
+    throw_pretty("Invalid argument: the upper bound is not equals / higher than the lower bound.")
+  }
+  if ((lower.array() == std::numeric_limits<Scalar>::infinity()).any()) {
+    throw_pretty("Invalid argument: the lower bound cannot containt a positive infinity value");
+  }
+  if ((upper.array() == -std::numeric_limits<Scalar>::infinity()).any()) {
+    throw_pretty("Invalid argument: the lower bound cannot containt a negative infinity value");
+  }
+
+  // Update the information related to the bounds
+  ng_ = 2 * ((upper - lower).array() != 0.).count() -
```

Comment:
I am not sure where you update the data buffers (Gx, Gu ... ) when you do this. Can you check, and at least mention it in a comment.

Reply:
In this PR, I am resizing the data inside the RK integrator. However, there are other cases that are missed here, which you can find them in the big PR. Note that I have unit-tests and working MPC (which changes ng online) using this latter PR.


Unfortunately, this PR represents a tiny part of the development and it is not possible to bring other features as they are are developed nonlineary. I will mark this conversation as resolved.

Reply:
Just to add more details. These are the type of modifications that you can see in the other PR:
  - function to easy modify dimension in the constraint manager: https://github.com/cmastalli/crocoddyl/blob/topic/preview-v2/include/crocoddyl/core/constraints/constraint-manager.hpp#L330-L348
  - use of this function in each diff action model, e.g.,https://github.com/cmastalli/crocoddyl/blob/topic/preview-v2/include/crocoddyl/multibody/actions/contact-fwddyn.hxx#L131
  - unit-test: https://github.com/cmastalli/crocoddyl/blob/topic/preview-v2/unittest/test_constraint_manager.cpp#L377
  - use the action dimensions to resize solver data online: https://github.com/cmastalli/crocoddyl/blob/topic/preview-v2/src/core/solvers/intro.cpp#L191-L216
  - virtual function to resize data in each solver: https://github.com/cmastalli/crocoddyl/blob/topic/preview-v2/include/crocoddyl/core/solver-base.hpp#L132

This last point also resolved our issue with different nu dimensions.

Reply:
I included these changes in the PR as they are an important part of the current API.